src/rubrix/labeling/text_classification/weak_labels.py

```diff
@@ -517,18 +617,343 @@ def change_mapping(self, label2int: Dict[str, int]):
-            self._annotation_array[annotation_masks[label]] = label2int[label]
+            self._annotation[annotation_masks[label]] = label2int[label]
 
         # update mapping dicts
         self._label2int = label2int.copy()
         self._int2label = {val: key for key, val in self._label2int.items()}
 
 
+class WeakMultiLabels(WeakLabelsBase):
+    """Computes the weak labels of a multi-label text classification dataset by applying a given list of rules.
+
+    Args:
+        dataset: Name of the dataset to which the rules will be applied.
+        rules: A list of rules (labeling functions). They must return a string, or ``None`` in case of abstention.
+            If None, we will use the rules of the dataset (Default).
+        ids: An optional list of record ids to filter the dataset before applying the rules.
+        query: An optional ElasticSearch query with the
+            `query string syntax <https://rubrix.readthedocs.io/en/stable/reference/webapp/search_records.html>`_
+            to filter the dataset before applying the rules.
+
+    Raises:
+        NoRulesFoundError: When you do not provide rules, and the dataset has no rules either.
+        DuplicatedRuleNameError: When you provided multiple rules with the same name.
+        SingleLabelError: When trying to get weak labels for a single-label text classification task.
+        NoRecordsFoundError: When the filtered dataset is empty.
+
+    Examples:
+        >>> # Get the 3 dimensional weak label matrix from a multi-label classification dataset with rules:
+        >>> weak_labels = WeakMultiLabels(dataset="my_dataset")
+        >>> weak_labels.matrix()
+        >>> weak_labels.summary()
+        >>>
+        >>> # Get the 3 dimensional weak label matrix from rules defined in Python:
+        >>> def awesome_rule(record: TextClassificationRecord) -> str:
+        ...     return ["Positive", "Slang"] if "next level" in record.inputs["text"] else None
+        >>> another_rule = Rule(query="amped OR psyched", label=["Positive", "Slang"])
+        >>> weak_labels = WeakMultiLabels(dataset="my_dataset", rules=[awesome_rule, another_rule])
+        >>> weak_labels.matrix()
+        >>> weak_labels.summary()
+    """
+
+    def __init__(
+        self,
+        dataset: str,
+        rules: Optional[List[Callable]] = None,
+        ids: Optional[List[Union[int, str]]] = None,
+        query: Optional[str] = None,
+    ):
+        super().__init__(dataset=dataset, rules=rules, ids=ids, query=query)
+
+        if not self._records[0].multi_label:
+            raise SingleLabelError(
+                "For single-label text classification, use the `rb.labeling.text_classification.WeakLabels` class."
+            )
+
+        # apply rules -> create the weak label matrix (3D), annotation (2D), label list
+        self._matrix, self._annotation, self._labels = self._apply_rules()
+
+    def _apply_rules(self) -> Tuple[np.ndarray, np.ndarray, List[str]]:
```

Comment:
Could be shared some logic from `_apply_rules` method for both, single and multi-label implementations?

Also, the single-label ones, accepts a `label2int` param and the multi-label don't. Don't we need this parameter in the multi label version?

Reply:
Just a summary of our quick call:

Haven't found a good way to share logics between both implementations for now, will leave this for a future iteration. 

> Also, the single-label ones, accepts a label2int param and the multi-label don't. Don't we need this parameter in the multi label version?

- The main idea of the `label2int` parameter was to facilitate the usage of the WeakLabels with Snorkel and FlyingSquid, but both can't deal with multi-label out of the box
- with our "binary" weak multi label matrix, a label2int is not really possible. maybe one could think about making the order of the labels configurable (order of the labels on the z axis), maybe this would be roughly equivalent, but not sure how useful this is in the end.